mpmath/libmp/libfmt.py

```diff
@@ -0,0 +1,424 @@
+from .libmpf import (to_digits_exp, stddigits, round_nearest, fzero, fnzero,
+                     finf, fninf, fnan)
+blog2 = 3.3219280948873626
+
+
+def round_digits(digits, dps, base, rounding=round_nearest):
+    '''
+    Returns the rounded digits, and the number of places the decimal point was
+    shifted.
+    '''
+
+    exponent = 0
+
+    if rounding == round_nearest:
+        rnd_digs = stddigits[(base//2 + base % 2):base]
+    else:
+        raise ValueError('Rounding: {} not implemented yet.')
+
+    round_up = False
+    if len(digits) > dps:
+        # The first digit after dps is a 5.
+        if digits[dps] == rnd_digs[0]:
+            for i in range(dps+1, len(digits)):
+                if digits[i] != '0':
+                    round_up = True
+                    break
+            if digits[dps-1] in stddigits[1:base:2]:
+                round_up = True
+
+    if not round_up:
+        digits = digits[:dps] + stddigits[int(digits[dps], base) - 1]
+
+    # Rounding up kills some instances of "...99999"
+    if len(digits) > dps and digits[dps] in rnd_digs:
+        digits = digits[:dps]
+        i = dps - 1
+        dig = stddigits[base-1]
+        while i >= 0 and digits[i] == dig:
+            i -= 1
+        if i >= 0:
+            digits = digits[:i] + stddigits[int(digits[i], base) + 1] + \
+                '0' * (dps - i - 1)
+        else:
+            digits = '1' + '0' * (dps - 1)
+            exponent += 1
+    else:
+        digits = digits[:dps]
+
+    return digits, exponent
+
+
+def calc_padding(nchars, width, align):
+    '''
+    Computes the left and right padding required to fill the required width,
+    according to how the string will be aligned.
+    '''
+    ntotal = max(nchars, width)
+
+    if align == '<':
+        lpad = 0
+        rpad = ntotal - nchars
+    elif align in ('>', '='):
+        lpad = ntotal - nchars
+        rpad = 0
+    elif align == '^':
+        lpad = (ntotal - nchars)//2
+        rpad = ntotal - nchars - lpad
+    else:
+        lpad = 0
+        rpad = 0
+
+    return (lpad, rpad)
+
+
+def read_format_spec(format_spec):
```

Comment:
I had copied over the code from `formatter_unicode.c` in CPython, but forgot the last section that checked this.
Anyways, I went ahead with your suggestion, the regex is a more elegant solution.
Also added a test so that `format(fp.mpf(1.1), "e ")` fails.